### PR TITLE
Auth and Biosolr for Solr 8.7

### PR DESCRIPTION
**Pull request by pcm32**
This PR adds:

- Authentication setup
- Use new package systems to upload a signed Biosolr JAR (which is not yet working well on being used in the schema, even though package creation, JAR and keys upload work well).

### CODE REVIEW

**Comment by pcm32** [comment]
Currently the package with biosolr gets created correctly, but only this final step fails when trying to use the package and the biosolr jar in the schema creation:

```
printf "\n\nCreate ontology expansion update processor"
HTTP_STATUS=$(curl $SOLR_AUTH -w "%{http_code}" -o >(cat >&3) -X POST -H 'Content-type:application/json' -d '{
  "add-updateprocessor":
  {
    "name": "'$CORE'_ontology_expansion",
    "class": "biosolr:uk.co.flax.biosolr.solr.update.processor.OntologyUpdateProcessorFactory",
    "annotationField": "ontology_annotation",
    "ontologyURI": "'$SCXA_ONTOLOGY'",
    "includeChildren": "false",
    "includeDescendants": "false"
  }
}' http://$HOST/solr/$CORE/config)
```

with this error:

```
Create ontology expansion update processor  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
0  5596  100  5248  100   348    174     11  0:00:31  0:00:30  0:00:01  1379{
  "responseHeader":{
    "status":500,
    "QTime":30022},
  "errorMessages":["1 out of 2 the property overlay to be of version 8 within 30 seconds! Failed cores: [http://172.19.0.3:8983/solr/scxa-analytics-v6_shard1_replica_n1/]\n"],
  "WARNING":"This response format is experimental.  It is likely to change in the future.",
  "error":{
    "metadata":[
      "error-class","org.apache.solr.common.SolrException",
      "root-error-class","org.apache.solr.common.SolrException"],
    "msg":"1 out of 2 the property overlay to be of version 8 within 30 seconds! Failed cores: [http://172.19.0.3:8983/solr/scxa-analytics-v6_shard1_replica_n1/]",
    "trace":"org.apache.solr.common.SolrException: 1 out of 2 the property overlay to be of version 8 within 30 seconds! Failed cores: [http://172.19.0.3:8983/solr/scxa-analytics-v6_shard1_replica_n1/]\n\tat org.apache.solr.handler.SolrConfigHandler.waitForAllReplicasState(SolrConfigHandler.java:848)\n\tat org.apache.solr.handler.SolrConfigHandler.access$400(SolrConfigHandler.java:106)\n\tat org.apache.solr.handler.SolrConfigHandler$Command.handleCommands(SolrConfigHandler.java:550)\n\tat org.apache.solr.handler.SolrConfigHandler$Command.handlePOST(SolrConfigHandler.java:382)\n\tat org.apache.solr.handler.SolrConfigHandler$Command.access$100(SolrConfigHandler.java:164)\n\tat org.apache.solr.handler.SolrConfigHandler.handleRequestBody(SolrConfigHandler.java:141)\n\tat org.apache.solr.handler.RequestHandlerBase.handleRequest(RequestHandlerBase.java:216)\n\tat org.apache.solr.core.SolrCore.execute(SolrCore.java:2637)\n\tat org.apache.solr.servlet.HttpSolrCall.execute(HttpSolrCall.java:794)\n\tat org.apache.solr.servlet.HttpSolrCall.call(HttpSolrCall.java:567)\n\tat org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:427)\n\tat org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:357)\n\tat org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:201)\n\tat org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)\n\tat org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:548)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)\n\tat org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)\n\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)\n\tat org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)\n\tat org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)\n\tat org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)\n\tat org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)\n\tat org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)\n\tat org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)\n\tat org.eclipse.jetty.server.handler.InetAccessHandler.handle(InetAccessHandler.java:177)\n\tat org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)\n\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)\n\tat org.eclipse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:322)\n\tat org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:763)\n\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)\n\tat org.eclipse.jetty.server.Server.handle(Server.java:516)\n\tat org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:400)\n\tat org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:645)\n\tat org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:392)\n\tat org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)\n\tat org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)\n\tat org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)\n\tat org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)\n\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)\n\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)\n\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)\n\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)\n\tat org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)\n\tat java.base/java.lang.Thread.run(Unknown Source)\n",
    "code":500}}

```

See https://github.com/ebi-gene-expression-group/index-scxa/runs/6403392811?check_suite_focus=true#step:3:624

**Comment by pcm32** [comment]
Could you please try this branch's `bash run_tests_in_containers.sh` locally @alfonsomunozpomer ? for some reason the keys fail in the CI, but everything works for me locally...

**Comment by pcm32** [comment]
I had to restart the job and then it worked, I guess the self hosted runner was in some odd state.

**Comment by pcm32** [comment]
One think I learnt here is that bats gets dizzy when two tests in the same bat file have the same name, so one should avoid that.

**Review by alfonsomunozpomer — APPROVED** [review]
It all looks good and it’s a big step forward! The only thing that worries me is how to handle authentication with the SolrJ libraries, but there should be a way to handle it.

Just out of curiosity, is the Solr container actually using ZooKeeper? I couldn’t spot the argument where it’s set.

File: README.md [hunk 1]
```diff
@@ -24,13 +37,16 @@ create-scxa-analytics-collection.sh
 ```
 
 ## Enable BioSolr
-`scxa-analytics-v5` makes use of the [BioSolr plugin](https://github.com/ebi-gene-expression-group/BioSolr) to perform ontology expansion on document indexing. In order to enable BioSolr, there are 2 options:
+
+`scxa-analytics-v5` makes use of the [BioSolr plugin](https://github.com/ebi-gene-expression-group/BioSolr) to perform ontology expansion on document indexing. In order to enable BioSolr, there are 3 options:
```
Comment (alfonsomunozpomer) [inline README.md:41]: `scxa-analytics-v7`? :wink:

**Comment by pcm32** [comment]
I tried the auth with the web cli, I think both here and at index-gxa, using the simplest method that gives the auth to the whole tomcat process and that works well. There are are couple more more granular methods, but they require changes in the Java code.

File: README.md (hunk 1 above)
Comment (pcm32) [reply README.md:41]: Yeah, probably at least v6. We didn't change the schema here (to move to v7), but we could as well do if you prefer.

**Comment by pcm32** [comment]
>Just out of curiosity, is the Solr container actually using ZooKeeper? I couldn’t spot the argument where it’s set.

I understand it should work with https://github.com/ebi-gene-expression-group/index-scxa/blob/feature/authentication_enabled/run_tests_in_containers.sh#L34 and I suspect it is working since the initial security upload works, which is through ZK. Or do you think that we are missing something?

**Comment by pcm32** [comment]
The auth with the web cli works like this: https://github.com/ebi-gene-expression-group/index-gxa/blob/feature/solr_8.7_authentication/bin/update_coexpressions_cli.sh#L38

**Comment by alfonsomunozpomer** [comment]
> > Just out of curiosity, is the Solr container actually using ZooKeeper? I couldn’t spot the argument where it’s set.
> 
> I understand it should work with https://github.com/ebi-gene-expression-group/index-scxa/blob/feature/authentication_enabled/run_tests_in_containers.sh#L34 and I suspect it is working since the initial security upload works, which is through ZK. Or do you think that we are missing something?

Yes, I had totally forgotten that SolrCloud was the default and just setting `ZK_HOST` is enough. We do that also with Solr 7.1... my bad!

**Comment by pcm32** [comment]
I have confirmed that our solr changes work on both bulk and SC, so I will start merging these PRs.

**Pull request merged**